gen/src/main.rs

```diff
@@ -0,0 +1,130 @@
+extern crate getopts;
+
+extern crate futures;
+extern crate hyper;
+extern crate tokio_core;
+
+#[macro_use]
+extern crate serde_derive;
+extern crate serde;
+extern crate serde_yaml;
+
+#[macro_use]
+extern crate lazy_static;
+
+#[macro_use]
+extern crate matches;
+extern crate regex;
+
+mod download;
+mod generate;
+
+use std::collections::HashMap;
+use std::env;
+use std::fs::File;
+use std::io::BufReader;
+use std::path::Path;
+
+use download::download;
+use generate::generate_all;
+
+use getopts::Options;
+
+fn main() {
+    let mut opts = Options::new();
+
+    opts.optflag("h", "help", "Print this help menu.");
+    opts.optopt(
+        "u",
+        "unicode-version",
+        "The version of the unicode standard to download.  \
+         If omitted, the local cache will be used.",
+        "VERSION",
+    );
+    opts.optmulti(
+        "",
+        "crate",
+        "A crate to generate tables for.  \
+         This propogates through the tree: if you specify only `unic`, all crates will be built.",
+        "CRATE",
+    );
+
+    let matches = opts.parse(env::args().skip(1))
+        .unwrap_or_else(|e| panic!(e.to_string()));
+
+    if matches.opt_present("h") {
+        print_usage(&opts);
+        return;
+    }
+
+    matches.opt_str("u").map(|version| {
+        println!("Downloading Unicode resources version {}...", version);
+        download(version.as_str()).expect("Failed to download Unicode resources");
+        println!("Finished.");
+    });
+
+    let crates = expand_sub_crates(matches.opt_strs("crate").into_iter());
+
+    if crates.is_empty() {
+        if !matches.opt_present("u") {
+            print_usage(&opts);
+        }
+        return;
+    }
+
+    println!("Generating sources for {} crates...", crates.len());
+    generate_all(crates).expect("Failed to generate sources");
+    println!("Finished.");
+}
+
+/// Path to crate disassembly mapping (YAML).
+///
+/// Format:
+///
+/// ```yaml
+/// ---
+/// crate:
+///   - "crate-subcrate1"
+///   - "crate-subcrate2"
+/// crate-subcrate1:
+///   - "crate-subcrate1-subcrate"
+/// ```
+const CRATES: &'static str = "gen/config/crates.yaml";
+
+/// Take a crate name in the UNIC hierarchy and expand it into all subcrates.
+///
+/// Reads `config/crates.yaml` for the crate disassembly process.
+///
+/// Technically, a crate only needs to be in that mapping if it needs to have sources built.
+/// For consistency, all crates that need generated sources or not should be passed into the
+/// generation module, which will ignore crates it doesn't need to do anything for.
+fn expand_sub_crates<I>(list: I) -> Vec<String>
```

Comment:
Okay, this is exactly what I was talking about on the previous comment section. Looks like we're on the same page now.

Yeah, I agree that we should focus on download/generate actions per source, and they can be plain linear for each source, without any option to do a subset of files/tables. This would keep the CLI simpler, and maintains enforcement of updating all the data (from each source) at the same time.

So, my only feedback to the first proposal here would be:
* Let's drop the `--target`, since the value provided there is mandatory (otherwise --help happens).
* And let's not call them `target`, but `data source` or just `source`, because there's possibility of having multiple targets per source.